x-exception/src/X/Control/Monad/Catch.hs

```diff
@@ -43,23 +44,25 @@ bracketF a f g =
 --
 -- Exception and `Left` safe version of bracketEitherT.
 --
-bracketEitherT' :: MonadMask m => EitherT e m a -> (a -> EitherT e m c) -> (a -> EitherT e m b) -> EitherT e m b
+bracketEitherT'
+  :: forall m e a b c. MonadMask m
+  => EitherT e m a
+  -> (a -> EitherT e m c)
+  -> (a -> EitherT e m b)
+  -> EitherT e m b
 bracketEitherT' acquire release run =
-  EitherT $ bracketF
-    (runEitherT acquire)
-    (\r -> case r of
-      Left _ ->
-        -- Acquire failed, we have nothing to release
-        return . Right $ ()
-      Right r' ->
-        -- Acquire succeeded, we need to try and release
-        runEitherT (release r') >>= \x -> return $ case x of
-          Left err -> Left (Left err)
-          Right _ -> Right ())
-    (\r -> case r of
-      Left err ->
-        -- Acquire failed, we have nothing to run
-        return . Left $ err
-      Right r' ->
-        -- Acquire succeeded, we can do some work
-        runEitherT (run r'))
+  let
+    acquire' :: m (Either e a)
+    acquire' = runEitherT acquire
+
+    release' :: Either e a -> m (Either e c)
+    release' = either (return . Left) (runEitherT . release) -- if acquire' failed then pass through
```

Comment:
The main diff with what was going on before was that `release'` was equivalent to `release' = either (const . return . Right $ ()) (runEitherT . release)`  But this case is moot anyway since if the acquire fails in this case both the run and release will fail passing through the acquire error and the run error will take precedence (they are the same anyway)